Move normal() to DistributionTemplates

pbelevich opened the pull request:
The purpose of this PR is to move `normal`/`normal_`/`normal_out` to `native/DistributionTemplates.h`, `native/cpu/DistributionTemplates.h` and `native/cuda/DistributionTemplates.h` to make it reusable for custom RNG, see cpu_rng_test.cpp as an example of custom RNG.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#35167 Move normal() to DistributionTemplates**

Differential Revision: [D20588248](https://our.internmc.facebook.com/intern/diff/D20588248)

pbelevich committed:
Move normal() to DistributionTemplates

[ghstack-poisoned]

pbelevich committed:
Update on "Move normal() to DistributionTemplates"

The purpose of this PR is to move `normal`/`normal_`/`normal_out` to `native/DistributionTemplates.h`, `native/cpu/DistributionTemplates.h` and `native/cuda/DistributionTemplates.h` to make it reusable for custom RNG, see cpu_rng_test.cpp as an example of custom RNG.




[ghstack-poisoned]

pbelevich committed:
Update on "Move normal() to DistributionTemplates"

The purpose of this PR is to move `normal`/`normal_`/`normal_out` to `native/DistributionTemplates.h`, `native/cpu/DistributionTemplates.h` and `native/cuda/DistributionTemplates.h` to make it reusable for custom RNG, see cpu_rng_test.cpp as an example of custom RNG.




[ghstack-poisoned]

pbelevich committed:
Update on "Move normal() to DistributionTemplates"

The purpose of this PR is to move `normal`/`normal_`/`normal_out` to `native/DistributionTemplates.h`, `native/cpu/DistributionTemplates.h` and `native/cuda/DistributionTemplates.h` to make it reusable for custom RNG, see cpu_rng_test.cpp as an example of custom RNG.




[ghstack-poisoned]

pbelevich committed:
Update on "Move normal() to DistributionTemplates"

The purpose of this PR is to move `normal`/`normal_`/`normal_out` to `native/DistributionTemplates.h`, `native/cpu/DistributionTemplates.h` and `native/cuda/DistributionTemplates.h` to make it reusable for custom RNG, see cpu_rng_test.cpp as an example of custom RNG.




[ghstack-poisoned]

pbelevich committed:
Update on "Move normal() to DistributionTemplates"

The purpose of this PR is to move `normal`/`normal_`/`normal_out` to `native/DistributionTemplates.h`, `native/cpu/DistributionTemplates.h` and `native/cuda/DistributionTemplates.h` to make it reusable for custom RNG, see cpu_rng_test.cpp as an example of custom RNG.




[ghstack-poisoned]

pbelevich committed:
Update on "Move normal() to DistributionTemplates"

The purpose of this PR is to move `normal`/`normal_`/`normal_out` to `native/DistributionTemplates.h`, `native/cpu/DistributionTemplates.h` and `native/cuda/DistributionTemplates.h` to make it reusable for custom RNG, see cpu_rng_test.cpp as an example of custom RNG.


Differential Revision: [D20588248](https://our.internmc.facebook.com/intern/diff/D20588248)

[ghstack-poisoned]

pbelevich committed:
Update on "Move normal() to DistributionTemplates"

The purpose of this PR is to move `normal`/`normal_`/`normal_out` to `native/DistributionTemplates.h`, `native/cpu/DistributionTemplates.h` and `native/cuda/DistributionTemplates.h` to make it reusable for custom RNG, see cpu_rng_test.cpp as an example of custom RNG.


Differential Revision: [D20588248](https://our.internmc.facebook.com/intern/diff/D20588248)

[ghstack-poisoned]

pbelevich committed:
Update on "Move normal() to DistributionTemplates"

The purpose of this PR is to move `normal`/`normal_`/`normal_out` to `native/DistributionTemplates.h`, `native/cpu/DistributionTemplates.h` and `native/cuda/DistributionTemplates.h` to make it reusable for custom RNG, see cpu_rng_test.cpp as an example of custom RNG.


Differential Revision: [D20588248](https://our.internmc.facebook.com/intern/diff/D20588248)

[ghstack-poisoned]

pbelevich committed:
Update on "Move normal() to DistributionTemplates"

The purpose of this PR is to move `normal`/`normal_`/`normal_out` to `native/DistributionTemplates.h`, `native/cpu/DistributionTemplates.h` and `native/cuda/DistributionTemplates.h` to make it reusable for custom RNG, see cpu_rng_test.cpp as an example of custom RNG.


Differential Revision: [D20588248](https://our.internmc.facebook.com/intern/diff/D20588248)

[ghstack-poisoned]

pbelevich committed:
Update on "Move normal() to DistributionTemplates"

The purpose of this PR is to move `normal`/`normal_`/`normal_out` to `native/DistributionTemplates.h`, `native/cpu/DistributionTemplates.h` and `native/cuda/DistributionTemplates.h` to make it reusable for custom RNG, see cpu_rng_test.cpp as an example of custom RNG.


Differential Revision: [D20588248](https://our.internmc.facebook.com/intern/diff/D20588248)

[ghstack-poisoned]

pbelevich committed:
Update on "Move normal() to DistributionTemplates"

The purpose of this PR is to move `normal`/`normal_`/`normal_out` to `native/DistributionTemplates.h`, `native/cpu/DistributionTemplates.h` and `native/cuda/DistributionTemplates.h` to make it reusable for custom RNG, see cpu_rng_test.cpp as an example of custom RNG.


Differential Revision: [D20588248](https://our.internmc.facebook.com/intern/diff/D20588248)

[ghstack-poisoned]

pbelevich committed:
Update on "Move normal() to DistributionTemplates"

The purpose of this PR is to move `normal`/`normal_`/`normal_out` to `native/DistributionTemplates.h`, `native/cpu/DistributionTemplates.h` and `native/cuda/DistributionTemplates.h` to make it reusable for custom RNG, see cpu_rng_test.cpp as an example of custom RNG.


Differential Revision: [D20588248](https://our.internmc.facebook.com/intern/diff/D20588248)

[ghstack-poisoned]

pbelevich committed:
Update on "Move normal() to DistributionTemplates"

The purpose of this PR is to move `normal`/`normal_`/`normal_out` to `native/DistributionTemplates.h`, `native/cpu/DistributionTemplates.h` and `native/cuda/DistributionTemplates.h` to make it reusable for custom RNG, see cpu_rng_test.cpp as an example of custom RNG.


Differential Revision: [D20588248](https://our.internmc.facebook.com/intern/diff/D20588248)

[ghstack-poisoned]

pbelevich committed:
Update on "Move normal() to DistributionTemplates"

The purpose of this PR is to move `normal`/`normal_`/`normal_out` to `native/DistributionTemplates.h`, `native/cpu/DistributionTemplates.h` and `native/cuda/DistributionTemplates.h` to make it reusable for custom RNG, see cpu_rng_test.cpp as an example of custom RNG.


Differential Revision: [D20588248](https://our.internmc.facebook.com/intern/diff/D20588248)

[ghstack-poisoned]

pbelevich committed:
Update on "Move normal() to DistributionTemplates"

The purpose of this PR is to move `normal`/`normal_`/`normal_out` to `native/DistributionTemplates.h`, `native/cpu/DistributionTemplates.h` and `native/cuda/DistributionTemplates.h` to make it reusable for custom RNG, see cpu_rng_test.cpp as an example of custom RNG.


Differential Revision: [D20588248](https://our.internmc.facebook.com/intern/diff/D20588248)

[ghstack-poisoned]

pbelevich committed:
Update on "Move normal() to DistributionTemplates"

The purpose of this PR is to move `normal`/`normal_`/`normal_out` to `native/DistributionTemplates.h`, `native/cpu/DistributionTemplates.h` and `native/cuda/DistributionTemplates.h` to make it reusable for custom RNG, see cpu_rng_test.cpp as an example of custom RNG.


Differential Revision: [D20588248](https://our.internmc.facebook.com/intern/diff/D20588248)

[ghstack-poisoned]

pbelevich committed:
Update on "Move normal() to DistributionTemplates"

The purpose of this PR is to move `normal`/`normal_`/`normal_out` to `native/DistributionTemplates.h`, `native/cpu/DistributionTemplates.h` and `native/cuda/DistributionTemplates.h` to make it reusable for custom RNG, see cpu_rng_test.cpp as an example of custom RNG.


Differential Revision: [D20588248](https://our.internmc.facebook.com/intern/diff/D20588248)

[ghstack-poisoned]

pbelevich commented:
@ailzhang what can be cause XLA failures?
test_normal_xla_float64
test_randn_xla_float32
test_randn_xla_float64

ailzhang commented:
cc: @JackCaoG who lowered `normal` :D. I assume we will not match `randn` result given we now have XLA-specific normal? 
Also we should likely fix `test_normal_xla_float64`.

pbelevich committed:
Update on "Move normal() to DistributionTemplates"

The purpose of this PR is to move `normal`/`normal_`/`normal_out` to `native/DistributionTemplates.h`, `native/cpu/DistributionTemplates.h` and `native/cuda/DistributionTemplates.h` to make it reusable for custom RNG, see cpu_rng_test.cpp as an example of custom RNG.


Differential Revision: [D20588248](https://our.internmc.facebook.com/intern/diff/D20588248)

[ghstack-poisoned]

pbelevich committed:
Update on "Move normal() to DistributionTemplates"

The purpose of this PR is to move `normal`/`normal_`/`normal_out` to `native/DistributionTemplates.h`, `native/cpu/DistributionTemplates.h` and `native/cuda/DistributionTemplates.h` to make it reusable for custom RNG, see cpu_rng_test.cpp as an example of custom RNG.


Differential Revision: [D20588248](https://our.internmc.facebook.com/intern/diff/D20588248)

[ghstack-poisoned]

anjali411 commented on aten/src/ATen/native/cuda/Copy.cu:
```diff
@@ -61,7 +61,7 @@ void copy_device_to_device(TensorIterator& iter, bool non_blocking) {
     }
   } else {
-    AT_DISPATCH_ALL_TYPES_AND3(kHalf, kBool, kBFloat16, iter.dtype(0), "copy_", [&] {
+    AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND3(kHalf, kBool, kBFloat16, iter.dtype(0), "copy_", [&] {
```
this is fixed in this PR https://github.com/pytorch/pytorch/pull/35344

pbelevich replied:
removed

anjali411 reviewed (approved):
lgtm after fixing the copy_() change :) thanks for the awesome work!

pbelevich committed:
Update on "Move normal() to DistributionTemplates"

The purpose of this PR is to move `normal`/`normal_`/`normal_out` to `native/DistributionTemplates.h`, `native/cpu/DistributionTemplates.h` and `native/cuda/DistributionTemplates.h` to make it reusable for custom RNG, see cpu_rng_test.cpp as an example of custom RNG.


Differential Revision: [D20588248](https://our.internmc.facebook.com/intern/diff/D20588248)

[ghstack-poisoned]

pbelevich commented:
@ailzhang @JackCaoG should I wait for XLA change or land this PR?

ailzhang commented on test/test_torch.py:
```diff
@@ -9737,53 +9727,131 @@ def assert_tuple_empty(tup, dim):
-        self.assertEqual(r.mean(), 2, 0.2)
-        self.assertEqual(r.std(), 3, 0.2)
+    @dtypes(torch.float, torch.double, torch.complex64, torch.complex128)
+    @dtypesIfCUDA(torch.float, torch.double, torch.complex64, torch.complex128)
```
Curious: these two list are identical, is this intended?

pbelevich replied:
fixed, thanks!

pbelevich committed:
Update on "Move normal() to DistributionTemplates"

The purpose of this PR is to move `normal`/`normal_`/`normal_out` to `native/DistributionTemplates.h`, `native/cpu/DistributionTemplates.h` and `native/cuda/DistributionTemplates.h` to make it reusable for custom RNG, see cpu_rng_test.cpp as an example of custom RNG.


Differential Revision: [D20588248](https://our.internmc.facebook.com/intern/diff/D20588248)

[ghstack-poisoned]

JackCaoG commented:
Sorry, missed the email earlier. I think you can land this pr, xla will skip new tests for now. I will investigate it a bit more, most likely test issue.

ailzhang reviewed (approved):
Thanks for waiting! This PR is good to go when you're ready. 
[note to self]:
It's fine to remove device dispatch for `normal` here since we have derivatives defined for 
```
normal(Tensor, double)
normal(double, Tensor)
normal(Tensor, Tensor)
```
So that they're still overridable by XLA. 
`normal(double, double)` is a composite op defined in `TensorFactories.cpp` which points to `normal_` as leaf node.
